### PR TITLE
CORE-9382 Retries when CPKs not available

### PR DESCRIPTION
**Pull request by mbrkic-r3**
Implemented retries of processing `VerifyContractsRequest` when error occurred during retrieving CPKs required for Verification Sandbox. Redelivery state is tracked with `VerifyContractsRequestRedelivery` and redeliveries are scheduled by `RedeliveryScheduler`.

### CODE REVIEW

File: components/ledger/ledger-verification/src/main/kotlin/net/corda/ledger/verification/processor/impl/VerificationRequestProcessor.kt [hunk 1]
```diff
@@ -1,53 +1,105 @@
...
     private val requestHandler: VerificationRequestHandler,
-    private val responseFactory: ResponseFactory
-) : DurableProcessor<String, VerifyContractsRequest> {
+    private val externalEventResponseFactory: ExternalEventResponseFactory
+) : StateAndEventProcessor<String, VerifyContractsRequestRedelivery, VerifyContractsRequest> {
 
     private companion object {
         val log = contextLogger()
         const val MDC_EXTERNAL_EVENT_ID = "external_event_id"
+        const val MAX_REDELIVERIES = 3
```
Comment (mbrkic-r3) [inline components/ledger/ledger-verification/src/main/kotlin/net/corda/ledger/verification/processor/impl/VerificationRequestProcessor.kt:36]: What would be the appropriate values for number of retries and delay between retries?

**Review by lankydan — COMMENTED** [review]
I haven't looked too closely because I have a fundamental question first.

Do we need to implement retries on the verification processor when we already have retries available as part of the external event framework?

You can return a transient error as the response of the verification request, upon the flow processor receiving this response it will resend another event to the verification worker. From what I understand of this change, that functionality should suffice.

**Comment by mbrkic-r3** [comment]
I thought so as well and checked with @mnesbit and he said it needs to be implement this way.
@mnesbit could you please put more details here?

**Comment by lankydan** [comment]
> I thought so as well and checked with @mnesbit and he said it needs to be implement this way. @mnesbit could you please put more details here?

@mnesbit is away, so if we want to progress on this we need to make a decision among ourselves.

I'd need to understand/see the technical reasons for the split because right now I know we've already handled missing CPKs in this way using the external events code.

@driessamyn do you have any insight on this?

**Comment by owenstanford** [comment]
> 

I tend to agree with Dan, In this cases, with a single producer (which I believe is the case here?), it makes sense to re-use the producers retry logic. The flow code (which I assume this pattern is taken from) is a very different case with regards to input events & state management, which is why it needs to generates it's own retry events.

**Comment by JamesHR3** [comment]
Are there any cases where we're expecting to do transaction verification checks outside of flow (well, ledger) code? I could see us wanting to do this in some recovery scenarios potentially?

File: components/ledger/ledger-verification/src/main/kotlin/net/corda/ledger/verification/LedgerVerificationComponent.kt [hunk 2]
```diff
@@ -45,7 +46,8 @@ class LedgerVerificationComponent @Activate constructor(
     private val verificationRequestSubscriptionFactory: VerificationSubscriptionFactory
 ) : Lifecycle {
     private var configHandle: Resource? = null
-    private var verificationProcessorSubscription: Subscription<String, VerifyContractsRequest>? = null
+    private var verificationProcessorSubscription:
+            StateAndEventSubscription<String, VerifyContractsRequestRedelivery, VerifyContractsRequest>? = null
```
Comment (mnesbit) [inline components/ledger/ledger-verification/src/main/kotlin/net/corda/ledger/verification/LedgerVerificationComponent.kt:50]: Talk to @fowlerrr as I think this is old way of managing these and we have a 'resource' pattern

**Comment by mnesbit** [comment]
I think flow retry would be a problem if/when we implement on-demand CPK fetch for old CPK's on the back chain. Retry could keep going to different workers and each would potentially have to fetch the CPK. If we decide that all 'trusted' CPK's go onto the cluster's compacted topic then possibly this isn't true, but this feels a more powerful pattern.

**Comment by owenstanford** [comment]
The flow external events framework should reuse the same ID/Key in a retry (@lankydan correct me If I'm wrong here) so it should go back to the same worker, I don't think the worker stickiness would be different between the two types of retry.

**Comment by lankydan** [comment]
> The flow external events framework should reuse the same ID/Key in a retry (@lankydan correct me If I'm wrong here) so it should go back to the same worker, I don't think the worker stickiness would be different between the two types of retry.

Yes, that is correct.

File: components/ledger/ledger-verification/src/main/kotlin/net/corda/ledger/verification/LedgerVerificationComponent.kt (hunk 2 above)
Comment (fowlerrr) [reply components/ledger/ledger-verification/src/main/kotlin/net/corda/ledger/verification/LedgerVerificationComponent.kt:50]: Yep.  I have a bit in the wiki about it: https://github.com/corda/corda-runtime-os/wiki/Lifecycle-Management#using-managed-resources-to-automate-closing-of-resources

You should be able to find examples in the code as well but these Resources (including the config) should be handled by the `LifecycleCoordinator`.  It's a simple way for us to make sure we don't forget to close them at the end.  And it's a tiny bit easier to use.

Comment (fowlerrr) [reply components/ledger/ledger-verification/src/main/kotlin/net/corda/ledger/verification/LedgerVerificationComponent.kt:50]: So, for example, below at line 73 you would have:

```
coordinator.createManagedResource(CONFIG_HANDLE) {
    configurationReadService.registerComponentForUpdates(
        coordinator,
        setOf(BOOT_CONFIG, MESSAGING_CONFIG)
    )
}
```

`CONFIG_HANDLE` is a const string defined in your class for now

Comment (fowlerrr) [reply components/ledger/ledger-verification/src/main/kotlin/net/corda/ledger/verification/LedgerVerificationComponent.kt:50]: And similar for the `verificationProcessorSubscription`:

```
            is ConfigChangedEvent -> {
                coordinator.createManagedResource(SUBSCRIPTION) {
                    verificationRequestSubscriptionFactory.create(
                        event.config.getConfig(MESSAGING_CONFIG)
                    )
                }
                logger.debug("Starting LedgerVerificationComponent.")
                coordinator.getManagedResource(SUBSCRIPTION)?.start()
                coordinator.updateStatus(LifecycleStatus.UP)
            }
```

**Comment by mbrkic-r3** [comment]
This is not needed anymore as it was decided that retries will be implemented using transient exceptions.

**Pull request closed without merging**